lib/membrane_rtsp_plugin/source/connection_manager.ex

```diff
@@ -117,109 +160,175 @@ defmodule Membrane.RTSP.Source.ConnectionManager do
-  end
-
-  defp build_transport_header(%{transport: {:udp, port_range_start, port_range_end}}, media_id) do
-    rtp_port = port_range_start + media_id * 2
+  @spec keep_alive(State.t()) :: State.t()
+  defp keep_alive(state) do
+    Membrane.Logger.debug("Send GET_PARAMETER to keep session alive")
+    RTSP.get_parameter_no_response(state.rtsp_session)
 
-    if rtp_port + 1 > port_range_end do
-      {:error, :port_range_exceeded}
-    else
-      {:ok, [{"Transport", "RTP/AVP/UDP;unicast;client_port=#{rtp_port}-#{rtp_port + 1}"}]}
-    end
+    %{state | keep_alive_timer: start_keep_alive_timer(state)}
   end
 
+  @spec start_keep_alive_timer(State.t()) :: reference()
   defp start_keep_alive_timer(%{keep_alive_interval: interval}) do
     Process.send_after(self(), :keep_alive, interval |> Membrane.Time.as_milliseconds(:round))
   end
 
-  defp keep_alive(state) do
-    Membrane.Logger.debug("Send GET_PARAMETER to keep session alive")
-    RTSP.get_parameter_no_response(state.rtsp_session)
+  @spec setup_rtsp_connection_with_tcp(RTSP.t(), [track()]) ::
+          {:ok, tracks :: [track()]} | {:error, reason :: term()}
+  defp setup_rtsp_connection_with_tcp(rtsp_session, tracks) do
+    socket = RTSP.get_socket(rtsp_session)
 
-    %{state | keep_alive_timer: start_keep_alive_timer(state)}
+    tracks
+    |> Enum.with_index()
+    |> Enum.reduce_while({:ok, []}, fn {track, idx}, {:ok, set_up_tracks} ->
+      transport_header =
+        [{"Transport", "RTP/AVP/TCP;unicast;interleaved=#{idx * 2}-#{idx * 2 + 1}"}]
+
+      case RTSP.setup(rtsp_session, track.control_path, transport_header) do
+        {:ok, %{status: 200}} ->
+          {:cont, {:ok, [%{track | transport: {:tcp, socket}} | set_up_tracks]}}
+
+        error ->
+          Membrane.Logger.debug(
+            "ConnectionManager: Setting up RTSP connection failed: #{inspect(error)}"
+          )
+
+          {:halt, {:error, :setting_up_rtsp_connection_failed}}
+      end
+    end)
+  end
+
+  @spec setup_rtsp_connection_with_udp(
+          RTSP.t(),
+          :inet.port_number(),
+          :inet.port_number(),
+          [track()],
+          [track()]
+        ) :: {:ok, tracks :: [track()]} | {:error, reason :: term()}
+  defp setup_rtsp_connection_with_udp(
+         rtsp_session,
+         min_port,
+         max_port,
+         tracks,
+         set_up_tracks \\ []
+       )
+
+  defp setup_rtsp_connection_with_udp(_rtsp_session, _port, _max_port, [], set_up_tracks) do
+    {:ok, Enum.reverse(set_up_tracks)}
+  end
+
+  defp setup_rtsp_connection_with_udp(_rtsp_session, max_port, max_port, _tracks, _set_up_tracks) do
+    {:error, :port_range_exceeded}
+  end
+
+  defp setup_rtsp_connection_with_udp(rtsp_session, port, max_port, tracks, set_up_tracks) do
+    if port_taken?(port) or port_taken?(port + 1) do
+      setup_rtsp_connection_with_udp(rtsp_session, port + 1, max_port, tracks, set_up_tracks)
```

Comment:
1. the port range is exceeded when port is equal to max_port, since we need port + 1 for RTCP channel, I can make this more explicit
2. why +2? The idea behind +1 is that if only the first port is taken then this will suffice. Do you think it would be better if we did increased port by 1 if port is taken, and by 2 if port + 1 is taken?

Reply:
You're right, I misread it 👍